bin/reth/src/db/tui.rs

```diff
@@ -0,0 +1,184 @@
+    }
+
+    /// Move to the previous list selection
+    fn previous(&mut self) {
+        let i = match self.state.selected() {
+            Some(i) => {
+                if i == 0 {
+                    self.entries.len() - 1
+                } else {
+                    i - 1
+                }
+            }
+            None => 0,
+        };
+        self.state.select(Some(i));
+    }
+
+    /// Show the [DbListTUI] in the terminal.
+    pub(crate) fn show_tui(
+        entries: HashMap<T::Key, T::Value>,
+        start: usize,
+        total_entries: usize,
+    ) -> eyre::Result<()> {
+        // setup terminal
+        enable_raw_mode()?;
+        let mut stdout = io::stdout();
+        execute!(stdout, EnterAlternateScreen, EnableMouseCapture)?;
+        let backend = CrosstermBackend::new(stdout);
+        let mut terminal = Terminal::new(backend)?;
+
+        // create app and run it
+        let tick_rate = Duration::from_millis(250);
+        let mut app = DbListTUI::<T>::new(entries, start, total_entries);
+        app.state.select(Some(0));
+        let res = run(&mut terminal, app, tick_rate);
+
+        // restore terminal
+        disable_raw_mode()?;
+        execute!(terminal.backend_mut(), LeaveAlternateScreen, DisableMouseCapture)?;
+        terminal.show_cursor()?;
+
+        if let Err(err) = res {
+            error!("{:?}", err)
+        }
+
+        Ok(())
+    }
+}
+
+fn run<B: Backend, T: Table>(
+    terminal: &mut Terminal<B>,
+    mut app: DbListTUI<T>,
+    tick_rate: Duration,
+) -> io::Result<()> {
+    let mut last_tick = Instant::now();
+    loop {
+        terminal.draw(|f| ui(f, &mut app))?;
+
+        let timeout =
+            tick_rate.checked_sub(last_tick.elapsed()).unwrap_or_else(|| Duration::from_secs(0));
+        if crossterm::event::poll(timeout)? {
+            if let Event::Key(key) = event::read()? {
+                match key.code {
+                    KeyCode::Char('q') => return Ok(()),
+                    KeyCode::Down => app.next(),
+                    KeyCode::Up => app.previous(),
+                    _ => {}
+                }
+            }
+        }
+        if last_tick.elapsed() >= tick_rate {
+            last_tick = Instant::now();
+        }
+    }
+}
+
+fn ui<B: Backend, T: Table>(f: &mut Frame<'_, B>, app: &mut DbListTUI<T>) {
+    let outer_chunks = Layout::default()
+        .direction(Direction::Vertical)
+        .constraints([Constraint::Percentage(95), Constraint::Percentage(5)].as_ref())
+        .split(f.size());
+
+    // Columns
+    {
+        let inner_chunks = Layout::default()
+            .direction(Direction::Horizontal)
+            .constraints([Constraint::Percentage(50), Constraint::Percentage(50)])
+            .split(outer_chunks[0]);
+
+        let formatted_keys = app
+            .entries
+            .keys()
+            .enumerate()
+            .map(|(i, k)| ListItem::new(format!("[{}] - {k:?}", i + app.start)))
```

Comment:
`cargo r -- db list Transactions --start 94018  --len 10`

Interesting, so on transactions this is shown like this..any idea why? seems weird that the index does not line up with the actual key?

<img width="667" alt="image" src="https://user-images.githubusercontent.com/17802178/213048433-605c3e48-5558-4aca-898c-bf106f4492db.png">

Reply:
Oh interesting! I think this is because of [`HashMap::keys()`](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.keys) & [`HashMap::values()`](https://doc.rust-lang.org/std/collections/struct.HashMap.html#method.values) returning elements in arbitrary order, didn't think about that. Will fix it up later today.